algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i

```diff
@@ -174,72 +156,142 @@ services::Status KernelImplRBF<defaultDense, algorithmFPType, cpu>::computeInter
-                                              (DAAL_INT *)&nFeatures, &zero, dataR, (DAAL_INT *)&nVectors1);
-        }
-        for (size_t i = 0; i < nVectors1; i++)
-        {
-            for (size_t k = 0; k < i; k++)
+
+        services::Status retStat =
+            Blas<algorithmFPType, cpu>::xgemm_blocked(&trans, &notrans, (DAAL_INT *)&nVectors2, (DAAL_INT *)&nVectors1, (DAAL_INT *)&nFeatures, &one,
+                                                      a2, (DAAL_INT *)&nFeatures, a1, (DAAL_INT *)&nFeatures, &zero, r, (DAAL_INT *)&nVectors2);
+        if (!retStat) return retStat;
+
+        WriteRows<algorithmFPType, cpu> mtR(r, 0, nVectors1);
+        DAAL_CHECK_BLOCK_STATUS(mtR);
+        algorithmFPType * dataR = mtR.get();
+
+        const size_t blockSize = 256;
+        const size_t blockNum  = nVectors1 / blockSize + !!(nVectors1 % blockSize);
+
+        daal::threader_for(blockNum, blockNum, [&](const size_t iBlock) {
+            const size_t numRowsInBlock = (iBlock != blockNum - 1) ? blockSize : nVectors1 - iBlock * blockSize;
+            const size_t startRow       = iBlock * blockSize;
+            const size_t finishRow      = startRow + numRowsInBlock;
+
+            for (size_t i = startRow; i < finishRow; i++)
             {
-                dataR[i * nVectors1 + k] = coeff * (dataR[i * nVectors1 + i] + dataR[k * nVectors1 + k] - two * dataR[i * nVectors1 + k]);
+                PRAGMA_IVDEP
+                PRAGMA_VECTOR_ALWAYS
+                for (size_t k = 0; k < i; k++)
+                {
+                    dataR[i * nVectors1 + k] = coeff * (dataR[i * nVectors1 + i] + dataR[k * nVectors1 + k] - two * dataR[i * nVectors1 + k]);
+                }
             }
-        }
-        for (size_t i = 0; i < nVectors1; i++)
-        {
-            dataR[i * nVectors1 + i] = zero;
-            daal::internal::Math<algorithmFPType, cpu>::vExp(i + 1, dataR + i * nVectors1, dataR + i * nVectors1);
-        }
-        for (size_t i = 0; i < nVectors1; i++)
-        {
-            for (size_t k = i + 1; k < nVectors1; k++)
+        });
+
+        daal::threader_for(blockNum, blockNum, [&](const size_t iBlock) {
+            const size_t numRowsInBlock = (iBlock != blockNum - 1) ? blockSize : nVectors1 - iBlock * blockSize;
+            const size_t startRow       = iBlock * blockSize;
+            const size_t finishRow      = startRow + numRowsInBlock;
+
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t i = startRow; i < finishRow; i++)
             {
-                dataR[i * nVectors1 + k] = dataR[k * nVectors1 + i];
+                dataR[i * nVectors1 + i] = zero;
+                daal::internal::Math<algorithmFPType, cpu>::vExp(i + 1, dataR + i * nVectors1, dataR + i * nVectors1);
             }
-        }
+        });
+
+        daal::threader_for(blockNum, blockNum, [&](const size_t iBlock) {
```

Comment:
This and the previous threader_for can be merged together to have more cache-friendly implementation.

Reply:
I think it's not possible because all the computations must be finished before the last threader_for begins. Otherwise an element of the matrix may be copied before it's been computed.
Or if you meant that I should copy elements and run `vExp(...)` after, then we get 2 times more computations. So I think it's not the best idea